hca/dss/__init__.py

```diff
@@ -107,87 +111,185 @@ def download(self, bundle_uuid, replica, version="", dest_name="",
+        checksum_index = 0
+        path_pieces = []
+        for prefix_length in cls.DIRECTORY_NAME_LENGTHS:
+            path_pieces.append(checksum[checksum_index:(checksum_index + prefix_length)])
+            checksum_index += prefix_length
+        path_pieces.append(checksum)
+        return os.path.join(*path_pieces)
+
+    @classmethod
+    def _parse_manifest(cls, manifest):
+        with open(manifest) as f:
+            # unicode_literals is on so all strings are unicode. CSV wants a str so we need to jump through a hoop.
+            delimiter = '\t'.encode('ascii') if USING_PYTHON2 else '\t'
+            reader = csv.DictReader(f, delimiter=delimiter, quoting=csv.QUOTE_NONE)
+            return list(reader)
+
+    def download_manifest_v2(self, manifest, replica, num_retries=10, min_delay_seconds=0.25):
+        """
+        Process the given manifest file in TSV (tab-separated values) format and download the files referenced by it.
+        The files are downloaded in the version 2 format.
+
+        This download format will serve as the main storage format for downloaded files. If a user specifies a different
+        format for download (coming in the future) the files will first be downloaded in this format, then hard-linked
+        to the user's preferred format.
+
+        :param str manifest: path to a TSV (tab-separated values) file listing files to download
+        :param str replica: the replica to download from. The supported replicas are: `aws` for Amazon Web Services, and
+            `gcp` for Google Cloud Platform. [aws, gcp]
+        :param int num_retries: The initial quota of download failures to accept before exiting due to
+            failures. The number of retries increase and decrease as file chucks succeed and fail.
+        :param float min_delay_seconds: The minimum number of seconds to wait in between retries.
+
+        Process the given manifest file in TSV (tab-separated values) format and download the files
+        referenced by it.
+
+        Each row in the manifest represents one file in DSS. The manifest must have a header row. The header row
+        must declare the following columns:
+
+        * `bundle_uuid` - the UUID of the bundle containing the file in DSS.
+
+        * `bundle_version` - the version of the bundle containing the file in DSS.
+
+        * `file_name` - the name of the file as specified in the bundle.
+
+        The TSV may have additional columns. Those columns will be ignored. The ordering of the columns is
+        insignificant because the TSV is required to have a header row.
+        """
+        rows = self._parse_manifest(manifest)
+        errors = 0
+        for row in rows:
+            file_uuid = row['file_uuid']
+            sha = row['file_sha256']
+            path = self._file_path(sha)
+            version = row['file_version']
+            size = row['file_size']
+            # This first check avoids downloading the file unnecessarily
```

Comment:
I think this is obvious, especially considering the log message in the `then` branch.

Reply:
good point.